src/mlpack/methods/gmm/gmm_diag.hpp

```diff
@@ -0,0 +1,299 @@
+/**
+ * @author Kim SangYeon
+ * @file gmm_diag.hpp
+ *
+ * Defines a Diagonal Gaussian Mixture model and estimates the parameters
+ * of the model.
+ *
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+
+#ifndef MLPACK_METHODS_GMM_DIAG_GMM_HPP
+#define MLPACK_METHODS_GMM_DIAG_GMM_HPP
+
+#include <mlpack/prereqs.hpp>
+
+// This is the default fitting method class.
+#include "em_fit.hpp"
+
+namespace mlpack {
+namespace gmm /** Gaussian Mixture Models. */ {
+
+/**
+ * A Diagonal Gaussian Mixture Model. 
+ * This class uses maximum likelihood loss functions to estimate the parameters
+ * of the DiagonalGMM on a given dataset via the given fitting mechanism, 
+ * defined by the FittingType template parameter.  The DiagonalGMM can be
+ * trained using normal data, or data with probabilities of being
+ * from this GMM (see DiagonalGMM::Train() for more information).
+ * The DiagonalGMM is the same as GMM except for wrapping gmm_diag class.
+ *
+ * The Train() method uses a template type 'FittingType'.  The FittingType
+ * template class must provide a way for the DigonalGMM to train on data.
+ * It must provide the following two functions:
+ *
+ * @code
+ * void Estimate(const arma::mat& observations,
+ *               std::vector<distribution::GaussianDistribution>& dists,
+ *               arma::vec& weights);
+ *
+ * void Estimate(const arma::mat& observations,
+ *               const arma::vec& probabilities,
+ *               std::vector<distribution::GaussianDistribution>& dists,
+ *               arma::vec& weights);
+ * @endcode
+ *
+ * Example use:
+ *
+ * @code
+ * // Set up a mixture of 5 gaussians in a 4-dimensional space.
+ * DiagonalGMM g(5, 4);
+ *
+ * // Train the DiagonalGMM given the data observations, using the default
+ * // EM fitting mechanism.
+ * 
+ * g.Train(data);
+ *
+ * // Get the probability of 'observation' being observed from this
+ * // DiagoanlGMM.
+ * double probability = g.Probability(observation);
+ *
+ * // Get a random observation from the DiagonalGMM.
+ * arma::vec observation = g.Random();
+ * @endcode
+ */
+class DiagonalGMM
+{
+ private:
+  //! The number of Gaussians in the model.
+  size_t gaussians;
+  //! The dimensionality of the model.
+  size_t dimensionality;
+
+  //! Vector of Gaussians
+  std::vector<distribution::GaussianDistribution> dists;
```

Comment:
So there are some really nice optimizations and speedups that can be made here---we don't need to hold `distribution::GaussianDistribution` here, because that class assumes a full covariance (and thus computes the Cholesky decomposition for faster probability computations).  But since we have a diagonal covariance, we can create a separate `distribution::DiagonalGaussianDistribution` class (or something like that), which internally can hold an `arma::vec` for the mean and an `arma::vec` for the diagonal covariance entries.  It could even hold the inverse of each diagonal covariance entry, so as to make the `Probability()` computation faster.

I remember that there was some other work with GANs where a Gaussian distribution with diagonal covariance was needed, but if I am remembering right at that time it was chosen that we would put their diagonal-specific code in the NN code.  Maybe the creation of a new `DiagonalGaussianDistribution` class could let us refactor that code a bit. :+1:

I'm not tied to the name `DiagonalGaussianDistribution`, by the way, it's just an idea.

Reply:
Thank you for letting me know, I'm so happy to know that way to optimize our codes. I'll look into it. :)